contracts/contracts/strategies/FraxConvexStrategy.sol

```diff
@@ -0,0 +1,473 @@
+        require(decimals0 == decimals1, "Decimals do not match");
+        if (CURVE_POOL_ASSETS_COUNT == 3) {
+            require(decimals1 == decimals2, "Decimals do not match");
+        }
+
+        InitializableAbstractStrategy._initialize(
+            _rewardTokenAddresses,
+            _assets,
+            _pTokens
+        );
+        _approveBase();
+    }
+
+    /**
+     * @dev Connects this strategy to the Curve implementation functions.
+     * In this case its a Curve pool with:
+     * 1. two coins
+     * 2. remove_liquidity_imbalance that includes a receiver
+     * 3. calc_token_amount that includes fees
+     */
+    function getCurveFunctions()
+        internal
+        pure
+        override(BaseCurveStrategy, CurveTwoCoinFunctions)
+        returns (CurveFunctions memory)
+    {
+        return CurveTwoCoinFunctions.getCurveFunctions();
+    }
+
+    /**
+     * @dev Unlock all the Frax Staked Convex LP tokens if:
+     * - there is a lock and
+     * - the lock has expired
+     * @return unlockedAmount The amount of Frax Staked Convex LP tokens unlocked.
+     * This will be zero if there is no lock or the lock has not yet expired.
+     */
+    function _unlock() internal returns (uint256 unlockedAmount) {
+        // If a lock exists and it has expired
+        if (lockKey != NO_KEY && block.timestamp > unlockTimestamp) {
+            // Withdraw all the Frax Staked Convex LP tokens from the lock
+            // to this strategy contract.
+            // Have to withdraw all as we can't withdraw a partial amount.
+            // The third `claim_rewards` parameter will still claim rewards even
+            // when set to `false`. It will only not collect rewards if the
+            // Frax Convex Locking contract has collectRewardsOnWithdrawalPaused = true.
+            unlockedAmount = IFraxConvexLocking(fraxLocking).withdrawLocked(
+                lockKey,
+                address(this),
+                false
+            );
+
+            emit Unlock(lockKey, unlockedAmount, unlockTimestamp);
+
+            // The previous withdraw deletes the old lock so reset
+            // slither-disable-next-line reentrancy-no-eth
+            lockKey = NO_KEY;
+        }
+        // else no lock exists or has not expired so nothing to do.
+    }
+
+    /**
+     * @dev Extends an existing lock for 7 days regardless of whether it has expired or not.
+     * Create a new lock if one doesn't exist and there is a positive target locked balance.
+     * Or add to the existing lock if the locked tokens is under the target locked balance.
+     * @param unlockedBalance The balance of the Frax Staked Convex LP tokens that are NOT locked.
+     * @param lockedBalance The balance of the Frax Staked Convex LP tokens that are locked.
+     * This includes any tokens in an expired lock.
+     */
+    function _lock(uint256 unlockedBalance, uint256 lockedBalance) internal {
```

Comment:
I think we could simplify the if logic in this somewhat.

Although the overall behavior of when viewed in how this is called works out, inside the function it looks like it could lock funds that we actually wanted to withdraw. Also we duplicate a write.

The first top level if overlaps with another if inside the lower level. Makes it a bit hard to reason about, since some parts of the process are in the earlier if, and some in a later if.

We should extend the time AFTER adding new funds. Otherwise we would be earning less.

Would be nice to emit an event with a zero amount if we extend only, that we we can externally track when our lock ends from events.

Might be cleaner like so

```
//math
// and nopes w/returns
//then
if(lockKey == NO_KEY){
 // new lock
} else {
  // add funds, if needed
  // always extend for max power
}
// event?
```

This way we will always leave this function with a max lock, unless we are waiting to funds to come out, in which case we would do nothing.

Since only the strategist will be moving funds to this, perhaps we can skip the min amount checks.

Reply:
Thanks @DanielVF. Excellent review as always

> Although the overall behavior of when viewed in how this is called works out, inside the function it looks like it could lock funds that we actually wanted to withdraw. Also we duplicate a write.

I originally had logic to withdraw expired locks even if the amount was less than the target locked balance. This made the unlocking logic more complicated as it depended whether it was being called from `withdraw` or `updateLock` as to what it needed to do. If the Strategist wants to withdraw from the expired lock amount, they will need to set the target lock balance lower, call `updateLock` and then call `withdraw`.

> The first top level if overlaps with another if inside the lower level. Makes it a bit hard to reason about, since some parts of the process are in the earlier if, and some in a later if.

Agree. I've gone for a new structure that is hopefully easier to follow.

> We should extend the time AFTER adding new funds. Otherwise we would be earning less.

Good pickup on the rewards being better with extending a lock after adding more funds. I've made this change.

> Would be nice to emit an event with a zero amount if we extend only, that we we can externally track when our lock ends from events.

Good idea on emitting a `Lock` event with zero amount if the lock was extended with no funds added. I've made this change.

> Might be cleaner like so

I've changed to this structure. There was a bug in the old implementation that kept extending the lock even when the locked balance was over the target locked balance.

> Since only the strategist will be moving funds to this, perhaps we can skip the min amount checks.

I've debated with myself a few times whether we need the min amount check. As you point out, its the Strategist who controls the funds in the strategy so I've dropped the min lock amount.

Frax's `lockLonger` will revert if the new unlock time is not extended. That is, it will revert if you lock to the same time. This is why I had a `unlockTimestamp < newUnlockTimestamp` check which would allow `updateLock` to be called twice in a block. But reverting a tx that is doing a second `updateLock` in a block is not a bad thing so I've removed this check.

Reply:
This function looks really good now!